### DISCUSSION TITLE
Replace SFINAE with static_assert

### DISCUSSION THREAD

**Discussion by MBalszun** [nlohmann/json#2890] [Ideas]
There are a couple of cases, where member function templates have constraints on the allowed template types (e.g. `detail::is_usable_as_key_type`). Those are usually expressed via sfinae, which will remove that template from the overload set if the constraint is not satisfied.

Imho that only makes sense, when the intention is to have multiple member function templates and activate one or the other depending on the parameter type. For things as the above mentioned `detail::is_usable_as_key_type` that have to be true for any overload, I think it is more sensible to have a static_assert inside the function, because it is a hard error to pass such a type.

That way the lib can give the user a clear error message that he tried to pass an illegal type instead of hoping that the compiler will produce a readable error message while listing all the possible overloads that fail to satisfy the constraint and the user trying to figure out, which was the desired overload.

**Comment by MBalszun** [comment]
Just realizing that this may be a problem specific to the string_view branch

**Reply by nlohmann** [reply]
We once had issues with static asserts, see #960. Since then, I am quite hesitant with them. But you're right - diagnosis is much better.

**Reply by MBalszun** [reply]
So this is a problem if someone else wants to sfinae based on whether it is ok to call a json function template with a given type, correct? That makes sense.
I didn't think about that scenario thanks.
